pep-0572.rst

```diff
@@ -169,25 +167,16 @@ in order to avoid ambiguities or user confusion:
   This rule is included to disallow excessively confusing code, and
   because parsing keyword arguments is complex enough already.
 
-- Assignment expressions (even parenthesized or occurring inside other
-  constructs) are prohibited in function default values.  For example,
-  the following examples are all invalid, even though the expressions
-  for the default values are valid in other contexts::
+- Assignment expressions are prohibited at the top level in function
```

Comment:
Does this mean assignment expressions are allowed in default values on nested functions? (Otherwise I'm not sure why "at the top level" is there.)

Reply:
Oh, horrible ambiguity! It's talking about the top level of the expression. The previous version of the PEP prohibited := *anywhere* in default values but that's awkward to enforce. How about this for a less ambiguous phrasing:

> A function argument default value cannot be an assignment expression (though it can *contain* one, e.g. parenthesized).

And I'll add an example with `# Valid, though not encouraged`.

Reply:
Oh wait, the "top level" terminology just echoes the other bullets in this section. I'll still doctor the phrasing but not as extensively.

Reply:
The current prohibitions in this section can be enforced easily within the grammar (there have to be separate productions for assignment expression and for an expression that's not an assignment anyway). But the prohibition on having assignments anywhere inside an expression would require either a complete duplication of all the expression-related grammar rules, or an explicit post-syntactic analysis step that looks for assignments and forbids them. Either of those would be pretty awkward.

In contrast, the rule elsewhere in the PEP forbidding an assignment target to clash with the loop control variable(s) in a comprehension can easily be implemented as part of the static analysis that's necessary anyway to determine the scope of variables.